### ISSUE TITLE
FR: Re-test only failing tests

### ISSUE THREAD

**Issue by krlmlr** [hadley/devtools#1654]
Often, I change some code, run the tests, just to notice failures. I then fix the code and re-run the tests until they pass. In this scenario I'm usually not interested in rerunning all tests, just the ones that failed.

Some GUIs for other environments offer a "re-run failing" button. For testthat, I need to use a filter, but I need to manually construct it, and I can't run the test in the "Build" pane of RStudio.

Can `devtools::test()` write a list with failing tests, and pick it up and use it as a filter on the next run of `devtools::test(filter = NULL)`? That file would be cleared once all tests are successful.

If we manage to output to the build pane (https://github.com/hadley/devtools/issues/1650#issuecomment-343465350), this could be done with an add-in too.

**Comment by jimhester** [comment]
I think this is a good idea, but feel like having a `retest()` or similar function name might be a clearer API than using `filter = NULL`.

There is also a few additional questions

1. Should it be a list of failed tests or test files? If the former how to handle test files that have additional setup code in the test file.
2. Which package should be responsible for storing the list of tests and the retest logic?
3. How should the failed tests be stored? The different reporters currently store the results in different ways.

**Comment by krlmlr** [comment]
If we just use `test()`, this works out of the box with the RStudio IDE, there the `test()` call seems hard-wired.

1. I'm fine with a list of test files, because the tests in a file are usually related.
1. devtools?
1. How about a reporter whose sole responsibility it is to capture files with at least one failing test?

I'm going to implement related functionality in brushthat first, we might be able to reuse code from there.

**Comment by hadley** [comment]
A somewhat related idea: if we adhere to the convention that every file in R/ should have a matching file in test/ we could simply have a “test current file” function, and expose it in an add in. I think that might be a lot simpler and still solve most common testing challenges.

**Comment by krlmlr** [comment]
This is much simpler, and a good approach for new packages, but for existing code? Also, I can imagine cases where a test uses functionality across several files.

(3.) FWIW, `devtools::test()` already returns a data frame with test and timing results, regardless of the reporter used.

**Comment by krlmlr** [comment]
Just found [testthis](https://github.com/s-fleck/testthis) by @s-fleck that offers functionality to run tests associated with a file.

**Comment by s-fleck** [comment]
@krlmlr Running just the tests in a the single associated test file works pretty well in my workflow. If you still find `retest()` would be useful - and to awkward for devtools - you can file a feature request for testthis and i'll try to figure something out.

**Comment by krlmlr** [comment]
I have implemented this functionality in brushthat -- by default only failing tests are re-run. To sum up:

- testthis already implements the add-in functionality to jump between source and test files, and to run individual tests
- `retest()` may be useful as a stand-alone functionality, maybe also in testthis
- maybe testthis and brushthat should really be just one package?

Closing, because this doesn't seem to be a devtools issue at this time.
